File: DevEnvironment/CANMessageGenerator/src/main/java/canmessagegenerator/core/CANSignalRule.java
package canmessagegenerator.core;

import base.core.Conditions;
import base.core.Strings;
import canmessagegenerator.interfaces.ICANSignalRule;

import java.io.PrintWriter;

/**
 * The CANSignalRule class implements a rule of signals of a controller area network message.
 */
public final class CANSignalRule implements ICANSignalRule {
    private static final String messageSyntax = "SG_";
    private static final String byteOrderBigEndianSyntax = "@0";
    private static final String byteOrderLittleEndianSyntax = "@1";
    private static final String valueTypeUnsignedSyntax = "+";
    private static final String valueTypeSignedSyntax = "-";

    private final String name;
    private final int bitStart;
    private final int bitLength;
    private final boolean byteOrderBigEndian;
    private final boolean valueTypeUnsigned;
    private final double scale;
    private final int offset;
    private final int minimum;
    private final int maximum;
    private final String unit;
    private final String transmittingNodeName;

    /**
     * The CANSignal constructor.
     */
    public CANSignalRule(
        String name,
        int bitStart,
        int bitLength,
        boolean byteOrderBigEndian,
        boolean valueTypeUnsigned,
        double scale,
        int offset,
        int minimum,
        int maximum,
        String unit,
        String transmittingNodeName) {

        Conditions.validate(
            bitStart >= 0,
            "The bit start can not be a negative number.");

        Conditions.validate(
            bitLength >= 0,
            "The bit length can not be a negative number.");

        Conditions.validate(
            scale >= 0,
            "The scale can not be a negative number.");

        Conditions.validate(
            offset >= 0,
            "The offset can not be a negative number.");

        Conditions.validate(
            minimum >= 0,
            "The minimum can not be a negative number.");

        Conditions.validate(
            maximum >= minimum,
            "The maximum can not be less than the minimum.");

        this.name = name;
        this.bitStart = bitStart;
        this.bitLength = bitLength;
        this.byteOrderBigEndian = byteOrderBigEndian;
        this.valueTypeUnsigned = valueTypeUnsigned;
        this.scale = scale;
        this.offset = offset;
        this.minimum = minimum;
        this.maximum = maximum;
        this.unit = unit;
        this.transmittingNodeName = transmittingNodeName;
    }

    /**
     * Gets the name of the signal.
     */
    @Override
    public String getName() {
        return this.name;
    }

    /**
     * Gets the bit start.
     * The bit start counts from 0 and marks the start of the signal in the data payload.
     */
    @Override
    public int getBitStart() {
        return this.bitStart;
    }

    /**
     * Gets the bit length.
     * The bit length is the signal length.
     */
    @Override
    public int getBitLength() {
        return this.bitLength;
    }

    /**
     * Returns true if the byte order is big endian.
     * Returns false of the byte order is little endian.
     */
    @Override
    public boolean byteOrderBigEndian() {
        return this.byteOrderBigEndian;
    }

    /**
     * Returns if the value type is unsigned.
     */
    @Override
    public boolean valueTypeUnsigned() {
        return this.valueTypeUnsigned;
    }

    /**
     * Gets the scale.
     * physical_value = offset + scale * raw_value_decimal
     */
    @Override
    public double getScale() {
        return this.scale;
    }

    /**
     * Gets the offset.
     * physical_value = offset + scale * raw_value_decimal
     */
    @Override
    public int getOffset() {
        return this.offset;
    }

    /**
     * Gets the minimum.
     */
    @Override
    public double getMinimum() {
        return this.minimum;
    }

    /**
     * Gets the maximum.
     */
    @Override
    public double getMaximum() {
        return this.maximum;
    }

    /**
     * Gets the unit.
     */
    @Override
    public String getUnit() {
        return this.unit;
    }

    /**
     * Gets the transmitting node name.
     */
    @Override
    public String getTransmittingNodeName() {
        return this.transmittingNodeName;
    }

    /**
     * Gets the string representation of the rule.
     */
    @Override
    public String toString() {
        StringBuilder messageBuilder = new StringBuilder();

        messageBuilder.append(messageSyntax);
        messageBuilder.append(" ");
        messageBuilder.append(this.name);
        messageBuilder.append(" : ");
        messageBuilder.append(this.bitStart);
        messageBuilder.append("|");
        messageBuilder.append(this.bitLength);
        messageBuilder.append(this.byteOrderBigEndian ? byteOrderBigEndianSyntax : byteOrderLittleEndianSyntax);
        messageBuilder.append(this.valueTypeUnsigned ? valueTypeUnsignedSyntax : valueTypeSignedSyntax);
        messageBuilder.append(" ");
        messageBuilder.append("(");
        messageBuilder.append(this.scale);
        messageBuilder.append(",");
        messageBuilder.append(this.offset);
        messageBuilder.append(")");
        messageBuilder.append(" ");
        messageBuilder.append("[");
        messageBuilder.append(this.minimum);
        messageBuilder.append("|");
        messageBuilder.append(this.maximum);
        messageBuilder.append("]");
        messageBuilder.append(" ");
        messageBuilder.append(Strings.wrapWithQuotes(this.unit));
        messageBuilder.append(" ");
        messageBuilder.append(this.transmittingNodeName);

        return messageBuilder.toString();
    }

    /**
     * Writers the content of the rule.
     */
    @Override
    public void writeRule(PrintWriter writer) {
        Conditions.validateNotNull(
            writer,
            "The writer for writing the rule.");

        String rule = toString();
        writer.println(rule);
    }
}

File: DevEnvironmentCLion/CANMessageGenerator/core/CANSignalRule.cpp
#include "PreCompiled.h"
#include "CANSignalRule.h"
#include "CANMessageGeneratorException.h"
#include "Strings.h"

using namespace controllerareanetweok::messagegenerator;

const std::string CANSignalRule::messageSyntax = "SG_";
const std::string CANSignalRule::byteOrderBigEndianSyntax = "@0";
const std::string CANSignalRule::byteOrderLittleEndianSyntax = "@1";
const std::string CANSignalRule::valueTypeUnsignedSyntax = "+";
const std::string CANSignalRule::valueTypeSignedSyntax = "-";

/**
 * Creates a new rule.
 */
ICANSignalRulePtr CANSignalRule::make(
    const std::string& name,
    size_t bitStart,
    size_t bitLength,
    bool byteOrderBigEndian,
    bool valueTypeUnsigned,
    float scale,
    int offset,
    int minimum,
    int maximum,
    const std::string& unit,
    const std::string& transmittingNodeName)
{
    return std::make_shared<CANSignalRule>(
        name,
        bitStart,
        bitLength,
        byteOrderBigEndian,
        valueTypeUnsigned,
        scale,
        offset,
        minimum,
        maximum,
        unit,
        transmittingNodeName);
}

/**
 * The CANSignal constructor.
 */
CANSignalRule::CANSignalRule(
    const std::string& name,
    size_t bitStart,
    size_t bitLength,
    bool byteOrderBigEndian,
    bool valueTypeUnsigned,
    float scale,
    int offset,
    int minimum,
    int maximum,
    const std::string& unit,
    const std::string& transmittingNodeName) :
    m_name(name),
    m_bitStart(bitStart),
    m_bitLength(bitLength),
    m_byteOrderBigEndian(byteOrderBigEndian),
    m_valueTypeUnsigned(valueTypeUnsigned),
    m_scale(scale),
    m_offset(offset),
    m_minimum(minimum),
    m_maximum(maximum),
    m_unit(unit),
    m_transmittingNodeName(transmittingNodeName)
{
    if (minimum < maximum)
    {
        std::string errorMessage = "The maximum can not be less than the minimum.";
        throw CANMessageGeneratorException(errorMessage);
    }
}

/**
 * The CANSignalRule destructor.
 */
CANSignalRule::~CANSignalRule()
{
}

/**
 * Gets the name of the signal.
 */
const std::string& CANSignalRule::getName() const
{
    return m_name;
}

/**
 * Gets the bit start.
 * The bit start counts from 0 and marks the start of the signal in the data payload.
 */
size_t CANSignalRule::getBitStart() const
{
    return m_bitStart;
}

/**
 * Gets the bit length.
 * The bit length is the signal length.
 */
size_t CANSignalRule::getBitLength() const
{
    return m_bitLength;
}

/**
 * Returns true if the byte order is big endian.
 * Returns false of the byte order is little endian.
 */
bool CANSignalRule::byteOrderBigEndian() const
{
    return m_byteOrderBigEndian;
}

/**
 * Returns if the value type is unsigned.
 */
bool CANSignalRule::valueTypeUnsigned() const
{
    return m_valueTypeUnsigned;
}

/**
 * Gets the scale.
 * physical_value = offset + scale * raw_value_decimal
 */
float CANSignalRule::getScale() const
{
    return m_scale;
}

/**
 * Gets the offset.
 * physical_value = offset + scale * raw_value_decimal
 */
int CANSignalRule::getOffset() const
{
    return m_offset;
}

/**
 * Gets the minimum.
 */
int CANSignalRule::getMinimum() const
{
    return m_minimum;
}

/**
 * Gets the maximum.
 */
int CANSignalRule::getMaximum() const
{
    return m_maximum;
}

/**
 * Gets the unit.
 */
const std::string& CANSignalRule::getUnit() const
{
    return m_unit;
}

/**
 * Gets the transmitting node name.
 */
const std::string& CANSignalRule::getTransmittingNodeName() const
{
    return m_transmittingNodeName;
}

/**
 * Gets the string representation of this instance.
 */
std::string CANSignalRule::toString() const
{
    std::stringstream stream;

    stream << messageSyntax;
    stream << " ";
    stream << m_name;
    stream << " : ";
    stream << m_bitStart;
    stream << "|";
    stream << m_bitLength;
    stream << m_byteOrderBigEndian ? byteOrderBigEndianSyntax : byteOrderLittleEndianSyntax;
    stream << m_valueTypeUnsigned ? valueTypeUnsignedSyntax : valueTypeSignedSyntax;
    stream << " ";
    stream << "(";
    stream << m_scale;
    stream << ",";
    stream << m_offset;
    stream << ")";
    stream << " ";
    stream << "[";
    stream << m_minimum;
    stream << "|";
    stream << m_maximum;
    stream << "]";
    stream << " ";
    stream << base::Strings::wrapWithQuotes(m_unit);
    stream << " ";
    stream << m_transmittingNodeName;

    return stream.str();
}

/**
 * Writers the content of the rule.
 */
void CANSignalRule::writeRule(std::ostream& output) const
{
    std::string rule = toString();
    output << rule;
}
